go/libraries/doltcore/merge/merge_prolly_indexes.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merge
+
+import (
+	"context"
+	"errors"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb/durable"
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/table/editor/creation"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/types"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+// indexEdit is an edit to a secondary index based on the primary row's key and value.
+// The members of tree.Diff have the following meanings:
+// - |Key| = the key to apply this edit on
+// - |From| = the previous value of this key (If nil, an insert is performed.)
+// - |To| = the new value of this key (If nil, a delete is performed.)
+//
+// It is invalid for |From| and |To| to be both be nil.
+type indexEdit interface {
```

Comment:
I'm not super excited about this interface if its only motivation is to use a common channel for `cellWiseMergeEdit` and `conflictEdit`. The allocation overhead of boxing these types as interfaces will likely outweigh the benefits of asynchronously updating secondary indexes. 

That being said, I don't think it's worth changing within this PR, we can take another pass at it in the future